File: ios/QimaWidget/Snapshot.swift
import Foundation

/// The data the app publishes for widgets (`WidgetSnapshot` in
/// lib/services/widget_snapshot.dart). Prices are canonical USD per troy
/// ounce (or per unit); every conversion below mirrors the app's
/// `PriceConverter` / `PortfolioHistory` rules so a widget and the app never
/// disagree on a number.
struct Snapshot: Decodable {
    /// Shared with the app (its entitlements). macOS uses a team-prefixed
    /// group, which every macOS version accepts without a consent prompt.
    #if os(macOS)
    static let appGroup = "ZS3A435WC2.com.devlabtechnologies.qima"
    #else
    static let appGroup = "group.com.devlabtechnologies.qima"
    #endif
    static let storageKey = "widget_snapshot"
    static let supportedVersion = 1

    let version: Int
    let updatedAt: Double
    let hideBalances: Bool
    let baseCurrency: String
    let currencies: [String: CurrencyInfo]
    let fxHistory: [String: [[Double]]]
    let cards: [Card]
    let instruments: [InstrumentData]
    let portfolio: Portfolio

    struct CurrencyInfo: Decodable {
        let rate: Double
        let symbol: String
        let suffix: Bool
    }

    struct Card: Decodable {
        let id: String
        let instrumentID: String
        let currency: String
        let unit: String
        let karat: Int?
    }

    struct InstrumentData: Decodable {
        let id: String
        let symbol: String
        let nameKey: String
        let assetClass: String
        let units: [String]
        let karats: [Int]
        let latest: Double?
        let latestAt: Double?
        let series: [String: [[Double]]]
    }

    struct Portfolio: Decodable {
        let available: Bool
        let valueUSD: Double?
        let costUSD: Double?
        let holdings: [Holding]?
        let series: [String: [[Double]]]?
    }

    struct Holding: Decodable {
        let instrumentID: String
        let valueUSD: Double
        let costUSD: Double
    }

    /// The last snapshot the app wrote, or nil before the first one (or if
    /// it was written by an app version this extension can't read).
    static func load() -> Snapshot? {
        guard let json = UserDefaults(suiteName: appGroup)?.string(forKey: storageKey),
              let data = json.data(using: .utf8),
              let snapshot = try? JSONDecoder().decode(Snapshot.self, from: data),
              snapshot.version == supportedVersion
        else { return nil }
        return snapshot
    }

    func instrument(_ id: String) -> InstrumentData? {
        instruments.first { $0.id == id }
    }

    func card(_ id: String) -> Card? {
        cards.first { $0.id == id }
    }

    /// Live rate: display units per 1 USD.
    func liveRate(_ currency: String) -> Double? {
        currency == "USD" ? 1 : currencies[currency]?.rate
    }

    /// Rate at [time] (ms): USD is 1; otherwise the most recent history
    /// sample at or before the time (the earliest sample for older times);
    /// with no history, the live rate — `PriceConverter.points`' rule.
    func rate(_ currency: String, at time: Double) -> Double? {
        if currency == "USD" { return 1 }
        guard let samples = fxHistory[currency], let first = samples.first else {
            return liveRate(currency)
        }
        var chosen = first[1]
        for sample in samples {
            if sample[0] <= time { chosen = sample[1] } else { break }
        }
        return chosen
    }
}

enum PriceUnit: String {
    case troyOunce, gram, kilogram, each

    static let gramsPerTroyOunce = 31.1034768

    var multiplier: Double {
        switch self {
        case .troyOunce, .each: return 1
        case .gram: return 1 / Self.gramsPerTroyOunce
        case .kilogram: return 1000 * (1 / Self.gramsPerTroyOunce)
        }
    }

    /// Abbreviation shown after an amount, e.g. "g"; nil for per-unit prices.
    var abbreviation: String? {
        switch self {
        case .troyOunce: return L10n.key("unit.abbr.troyOunce")
        case .gram: return L10n.key("unit.abbr.gram")
        case .kilogram: return L10n.key("unit.abbr.kilogram")
        case .each: return nil
        }
    }
}

enum ChartWindow: String, CaseIterable {
    case day = "1D", week = "1W", month = "1M", quarter = "3M", year = "1Y", all = "ALL"
}

struct ChartPoint: Identifiable {
    let date: Date
    let value: Double
    var id: Date { date }
}

/// A range-scoped change, first point to last (spec §8.12: never an
/// all-time figure next to a bounded chart).
struct Change {
    let absolute: Double
    let fraction: Double
    var isUp: Bool { fraction >= 0 }

    init?(_ points: [ChartPoint]) {
        guard points.count >= 2, let first = points.first?.value, let last = points.last?.value, first != 0 else {
            return nil
        }
        absolute = last - first
        fraction = (last - first) / first
    }
}

File: android/app/src/main/kotlin/com/devlabtechnologies/qima/widget/Snapshot.kt
package com.devlabtechnologies.qima.widget

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject

/**
 * The data the app publishes for widgets (`WidgetSnapshot` in
 * `lib/services/widget_snapshot.dart`), Kotlin port of `Snapshot.swift`.
 * Prices are canonical USD per troy ounce (or per unit); every conversion
 * below mirrors the app's `PriceConverter` / `PortfolioHistory` rules so a
 * widget and the app never disagree on a number.
 */
class Snapshot(private val json: JSONObject) {

    companion object {
        const val STORAGE_KEY = "widget_snapshot"
        const val SUPPORTED_VERSION = 1

        /** SharedPreferences file `home_widget`'s Dart API writes to
         * (`HomeWidget.saveWidgetData`). `HomeWidgetPlugin.PREFERENCES`
         * holds the same literal but is `internal` to the plugin, so this
         * app module can't reference it directly. */
        private const val HOME_WIDGET_PREFS = "HomeWidgetPreferences"

        /** The last snapshot the app wrote, or null before the first one (or
         * if it was written by an app version this widget can't read). */
        fun load(context: Context): Snapshot? {
            val prefs = context.getSharedPreferences(HOME_WIDGET_PREFS, Context.MODE_PRIVATE)
            val raw = prefs.getString(STORAGE_KEY, null) ?: return null
            if (raw.isBlank()) return null
            return try {
                val obj = JSONObject(raw)
                if (obj.optInt("version", -1) != SUPPORTED_VERSION) null else Snapshot(obj)
            } catch (e: org.json.JSONException) {
                null
            }
        }
    }

    val hideBalances: Boolean get() = json.optBoolean("hideBalances", false)
    val baseCurrency: String get() = json.optString("baseCurrency", "USD")

    data class CurrencyInfo(val rate: Double, val symbol: String, val suffix: Boolean)

    private val currencies: Map<String, CurrencyInfo> by lazy {
        val obj = json.optJSONObject("currencies") ?: JSONObject()
        obj.keys().asSequence().associateWith { code ->
            val info = obj.getJSONObject(code)
            CurrencyInfo(info.optDouble("rate", 1.0), info.optString("symbol", code), info.optBoolean("suffix", false))
        }
    }

    fun currency(code: String): CurrencyInfo? = currencies[code]

    val currencyCodes: List<String> get() = currencies.keys.sorted()

    data class Card(val id: String, val instrumentID: String, val currency: String, val unit: String, val karat: Int?)

    val cards: List<Card> by lazy {
        val array = json.optJSONArray("cards") ?: JSONArray()
        (0 until array.length()).map { i ->
            val obj = array.getJSONObject(i)
            Card(
                id = obj.getString("id"),
                instrumentID = obj.getString("instrumentID"),
                currency = obj.getString("currency"),
                unit = obj.getString("unit"),
                karat = if (obj.isNull("karat")) null else obj.optInt("karat"),
            )
        }
    }

    fun card(id: String): Card? = cards.firstOrNull { it.id == id }

    data class InstrumentData(
        val id: String,
        val symbol: String,
        val nameKey: String,
        val assetClass: String,
        val units: List<String>,
        val karats: List<Int>,
        val latest: Double?,
        val latestAt: Double?,
        val series: Map<String, List<DoubleArray>>,
    )

    val instruments: List<InstrumentData> by lazy {
        val array = json.optJSONArray("instruments") ?: JSONArray()
        (0 until array.length()).map { i ->
            val obj = array.getJSONObject(i)
            val unitsArray = obj.optJSONArray("units") ?: JSONArray()
            val units = (0 until unitsArray.length()).map { unitsArray.getString(it) }
            val karatsArray = obj.optJSONArray("karats") ?: JSONArray()
            val karats = (0 until karatsArray.length()).map { karatsArray.getInt(it) }
            val seriesObj = obj.optJSONObject("series") ?: JSONObject()
            val series = seriesObj.keys().asSequence().associateWith { rangeKey ->
                val points = seriesObj.getJSONArray(rangeKey)
                (0 until points.length()).map { p ->
                    val point = points.getJSONArray(p)
                    doubleArrayOf(point.getDouble(0), point.getDouble(1))
                }
            }
            InstrumentData(
                id = obj.getString("id"),
                symbol = obj.optString("symbol", ""),
                nameKey = obj.optString("nameKey", ""),
                assetClass = obj.optString("assetClass", ""),
                units = units,
                karats = karats,
                latest = if (obj.isNull("latest")) null else obj.optDouble("latest"),
                latestAt = if (obj.isNull("latestAt")) null else obj.optDouble("latestAt"),
                series = series,
            )
        }
    }

    fun instrument(id: String): InstrumentData? = instruments.firstOrNull { it.id == id }

    data class Holding(val instrumentID: String, val valueUSD: Double, val costUSD: Double)

    data class Portfolio(
        val available: Boolean,
        val valueUSD: Double?,
        val costUSD: Double?,
        val holdings: List<Holding>?,
        val series: Map<String, List<DoubleArray>>?,
    )

    val portfolio: Portfolio by lazy {
        val obj = json.optJSONObject("portfolio") ?: JSONObject()
        val available = obj.optBoolean("available", false)
        if (!available) {
            return@lazy Portfolio(false, null, null, null, null)
        }
        val holdingsArray = obj.optJSONArray("holdings") ?: JSONArray()
        val holdings = (0 until holdingsArray.length()).map { i ->
            val h = holdingsArray.getJSONObject(i)
            Holding(h.getString("instrumentID"), h.getDouble("valueUSD"), h.getDouble("costUSD"))
        }
        val seriesObj = obj.optJSONObject("series") ?: JSONObject()
        val series = seriesObj.keys().asSequence().associateWith { rangeKey ->
            val points = seriesObj.getJSONArray(rangeKey)
            (0 until points.length()).map { p ->
                val point = points.getJSONArray(p)
                doubleArrayOf(point.getDouble(0), point.getDouble(1), point.getDouble(2))
            }
        }
        Portfolio(
            available = true,
            valueUSD = obj.optDouble("valueUSD"),
            costUSD = obj.optDouble("costUSD"),
            holdings = holdings,
            series = series,
        )
    }

    /** Live rate: display units per 1 USD. */
    fun liveRate(currency: String): Double? = if (currency == "USD") 1.0 else currencies[currency]?.rate

    private val fxHistory: Map<String, List<DoubleArray>> by lazy {
        val obj = json.optJSONObject("fxHistory") ?: JSONObject()
        obj.keys().asSequence().associateWith { code ->
            val points = obj.getJSONArray(code)
            (0 until points.length()).map { p ->
                val point = points.getJSONArray(p)
                doubleArrayOf(point.getDouble(0), point.getDouble(1))
            }
        }
    }

    /**
     * Rate at [time] (ms): USD is 1; otherwise the most recent history
     * sample at or before the time (the earliest sample for older times);
     * with no history, the live rate -- `PriceConverter.points`' rule.
     */
    fun rate(currency: String, time: Double): Double? {
        if (currency == "USD") return 1.0
        val samples = fxHistory[currency]
        if (samples.isNullOrEmpty()) return liveRate(currency)
        var chosen = samples.first()[1]
        for (sample in samples) {
            if (sample[0] <= time) chosen = sample[1] else break
        }
        return chosen
    }
}

/** Unit a price can be expressed in, matching `PriceUnit` (asset.dart / Snapshot.swift). */
enum class PriceUnit(val rawValue: String) {
    TROY_OUNCE("troyOunce"),
    GRAM("gram"),
    KILOGRAM("kilogram"),
    EACH("each");

    companion object {
        const val GRAMS_PER_TROY_OUNCE = 31.1034768

        fun fromRawValue(value: String?): PriceUnit? = values().firstOrNull { it.rawValue == value }
    }

    val multiplier: Double
        get() = when (this) {
            TROY_OUNCE, EACH -> 1.0
            GRAM -> 1.0 / GRAMS_PER_TROY_OUNCE
            KILOGRAM -> 1000.0 * (1.0 / GRAMS_PER_TROY_OUNCE)
        }
}

/** Selectable chart windows, matching `WidgetSnapshot.ranges` / `ChartWindow`. */
enum class ChartWindow(val id: String) {
    DAY("1D"), WEEK("1W"), MONTH("1M"), QUARTER("3M"), YEAR("1Y"), ALL("ALL");
}

data class ChartPoint(val timeMillis: Double, val value: Double)

/**
 * A range-scoped change, first point to last (spec §8.12: never an all-time
 * figure next to a bounded chart).
 */
class Change private constructor(val absolute: Double, val fraction: Double) {
    val isUp: Boolean get() = fraction >= 0

    companion object {
        fun from(points: List<ChartPoint>): Change? {
            if (points.size < 2) return null
            val first = points.first().value
            val last = points.last().value
            if (first == 0.0) return null
            return Change(last - first, (last - first) / first)
        }
    }
}
